### PR TITLE
Implement minify-builtins plugin

### LINKED ISSUE #363: Create aliases for built in functions
I mean, if I use multiple times `Math.floor`, it would make sense to assign it to a variable and use the variable. 

```
function(x, y,z) {
  return { Math.floor(x), Math.floor(y), Math.floor(z)  };
}
```

To

```
function(x, y, z) {
  let a = Math.floor;
  return { a(x), a(y), a(z) };
}
```

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
fixes #363

### CODE REVIEW

**Comment by boopathi** [comment]
Thoughts: I think we could also help constant-propagation by evaluating the expressions if the arguments are constants ?

```js
let a = 1.8, b = 2.5;
let x = Math.floor(Math.max(a, b));
foo(x);
```

could be transformed to

```js
foo(2);
```

**Comment by vigneshshanmugam** [comment]
@boopathi I like the idea in general, I believe it shouldn't cause any issues.. 

Shall i create a new issue for that and fix it in a new PR?

**Comment by boopathi** [comment]
Part of this PR would be great.

**Comment by vigneshshanmugam** [comment]
Cool then.. will give it a try..

**Comment by vigneshshanmugam** [comment]
Handled the optimisations.

File: packages/babel-plugin-transform-built-ins/src/builtins.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
+module.exports = {
```
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/src/builtins.js]: The way babel does this is to maintain only `[String, Number, Math]` 

https://github.com/babel/babel/blob/f8ffe03/packages/babel-traverse/src/path/evaluation.js#L5 

So, I think it's relatively safe to assume that `Math.x` will be a side-effect free evaluation except `Math.random`

File: packages/babel-plugin-transform-built-ins/src/index.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        CallExpression: {
+          exit(path) {
+            const callee = path.get("callee");
+            if (!callee.isMemberExpression()) {
+              return;
+            }
+
+            const expName = getExpressionName(callee);
+            if (isBuiltin(expName)) {
+              const result = evaluate(path);
```
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/src/index.js]: we have to deopt when we have side-effecty arguments 

for example,

```js
Math.floor((foo(), 1));
```

File: packages/babel-plugin-transform-built-ins/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+# babel-plugin-transform-built-ins
```
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/README.md]: should we just call it `babel-plugin-minify-builtins`

File: packages/babel-plugin-transform-built-ins/src/index.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
+"use strict";
+
+const { methods, properties } = require("./builtins");
+const evaluate = require("babel-helper-evaluate-path");
+
+
+
```
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/src/index.js]: empty lines

File: packages/babel-plugin-transform-built-ins/src/index.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        },
+        CallExpression: {
+          exit(path) {
+            const callee = path.get("callee");
+            if (!callee.isMemberExpression()) {
+              return;
+            }
+
+            const expName = getExpressionName(callee);
+            if (isBuiltin(expName)) {
```
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/src/index.js]: should check if it's a computed property

Will fail for `Math[max]()` will assume it as `Math.max()`.

File: packages/babel-plugin-transform-built-ins/src/index.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    },
+  };
+
+  function memberToString(memberExpr) {
+    const {object, property} = memberExpr.node;
+    let result = "";
+
+    if (t.isIdentifier(object)) result += object.name;
+    if (t.isMemberExpression(object)) result += memberToString(object);
+    if (t.isIdentifier(property)) result += property.name;
```
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/src/index.js]: should check for computed properties. same as above.

File: packages/babel-plugin-transform-built-ins/src/builtins.js (hunk 1 above)
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-built-ins/src/builtins.js]: Ahh good one.. What about Date.x?

File: packages/babel-plugin-transform-built-ins/README.md (hunk 3 above)
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-built-ins/README.md]: Hmmm.. dono really..

File: packages/babel-plugin-transform-built-ins/src/builtins.js (hunk 1 above)
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/src/builtins.js]: We can chuck the Date for now.

File: packages/babel-plugin-transform-built-ins/README.md (hunk 3 above)
Comment (boopathi) [inline packages/babel-plugin-transform-built-ins/README.md]: Let's just change it to `babel-plugin-minify-builtins`

File: packages/babel-plugin-transform-built-ins/src/builtins.js (hunk 1 above)
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-built-ins/src/builtins.js]: Cool.. then will update

File: packages/babel-plugin-transform-built-ins/README.md (hunk 3 above)
Comment (vigneshshanmugam) [inline packages/babel-plugin-transform-built-ins/README.md]: Alright

**Comment by vigneshshanmugam** [comment]
@boopathi Addressed all the issues.

**Comment by FezVrasta** [comment]
Is this already available in `0.0.11`?

**Comment by vigneshshanmugam** [comment]
Nope.. Not released yet.

**Pull request merged**